### ISSUE TITLE
tensorflow.GraphDef was modified concurrently during serialization.

### ISSUE THREAD

**Issue by mikeun** [tensorpack/tensorpack#781]
```
[0605 17:29:01 @summary.py:38] Maintain moving average summary of 6 tensors in collection MOVING_SUMMARY_OPS.
[0605 17:29:01 @summary.py:75] Summarizing collection 'summaries' of size 182.
[0605 17:29:01 @graph.py:91] Applying collection UPDATE_OPS of 8 ops.
[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/message_lite.cc:68] CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.
```
Unexpected Problems / Potential Bugs
I made the Trainer as your GAN
```
class AnalyzeTrainer(TowerTrainer):
    def __init__(self, input, model, command=None):
        super(AnalyzeTrainer, self).__init__()
        assert isinstance(model, AnalyzeModelDesc), model
        a = GetConfig()
        inputs_desc = model.get_inputs_desc()
        cbs = input.setup(inputs_desc)
        self.register_callback(cbs)
        # Build the graph
        self.tower_func = TowerFuncWrapper(model.build_graph, inputs_desc)
        with TowerContext('', is_training=True):
            self.tower_func(*input.get_input_tensors())
        opt = model.get_optimizer()

        with tf.name_scope('optimize'):
            if command == a.json_params['analyze_command']:
                minimize_class = opt.minimize(
                    model.analyze_class_cost,
                    var_list=model.varlist_analyze,
                    name='minimize_analyze_class_cost'
                )
                with tf.control_dependencies([minimize_class]):
                    train_op = opt.minimize(
                        model.analyze_pdf_cost,
                        var_list=model.varlist_analyze,
                        name='minimize_analyze_pdf_cost'
                    )
            elif command in [
                    a.json_params['retrain_command'],
                    a.json_params['train_new_command']
            ]:
                train_op = opt.minimize(
                    model.class_cost,
                    var_list=model.varlist_train,
                    name='minimize_class_cost'
                )
            elif command == a.json_params['train_pdf_command']:
                train_op = opt.minimize(
                    model.epdf_cost,
                    var_list=model.varlist_pdf,
                    name='minimize_pdf_cost'
                )
            else:
                assert False, " <+> there is not command \"{}\"".format(command)
        self.train_op = train_op
```

also ModelDesc:

```
class AnalyzeModelDesc(ModelDescBase):
    def collect_variables(
            self,
            analyze_collection='analyze_variables',
            pdf_collection='pdf_variables',
            train_collection='train_variables',
            class_cost=None,
            epdf_cost=None,
            analyze_class_cost=None,
            analyze_pdf_cost=None
    ):
        self.varlist_analyze = tf.get_collection(analyze_collection)
        self.varlist_pdf = tf.get_collection(pdf_collection)
        self.varlist_train = tf.get_collection(train_collection)
        self.class_cost = class_cost
        self.epdf_cost = epdf_cost
        self.analyze_class_cost = analyze_class_cost
        self.analyze_pdf_cost = analyze_pdf_cost

    def build_graph(self, *inputs):
        """
        Have to build one tower and set the following attributes:
        collections: analyze_variables, pdf_variables, train_variables.
        variables: self.class_cost, self.epdf_cost, self.analyze_cost
        """
        pass

    @memoized
    def get_optimizer(self):
        return self.optimizer()

    @abstractmethod
    def optimizer(self):
        pass
```

error after:

```
    trainer = Analyze.AnalyzeTrainer(data, setup_model.Model(), command=command)
    trainer.train_with_defaults(**config)
```

environment:
	+ Python 3.6.2 |Intel Corporation| (default, Aug 15 2017, 16:14:26) 
	+ TF version: v1.8.0-2154-ga2e1b4dcbd' 1.8.0
	+ Tensorpack version: 0.8.5

**Comment by ppwwyyxx** [comment]
I don't know how to reproduce this error and this is some error I've never seen before. Could be a TensorFlow bug as well. Having a piece of runable code that can reproduce the error would be awesome. Apart from that, some information that may make it easier to debug:
1. Have you seen the log "creating the session...." ? Usually this line of log comes after "Applying collection UPDATE_OPS of 8 ops.". Knowing where it crashes can help a lot.
2. Does your code consistently crash or crash by chance? What about a different version of TensorFlow?
3. What does "config" look like?
4. If you add some weird stuff to TF collections (i.e., things that are not ops or tensors), try removing them

**Comment by mikeun** [comment]
1. here is log of error:

```
[0605 18:05:11 @graph.py:91] Applying collection UPDATE_OPS of 8 ops.
[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/message_lite.cc:68] CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)`
```

there is no "creating session ...."

2. My program is crashed consistently. I have not any other version of TF.

3. config in this case is:

```
        dataset_train, dataset_test, _ = get_data(command=command)
        steps_per_epoch = dataset_train.size()
        dataset_train = StagingInput(QueueInput(dataset_train))
        dataset_test = QueueInput(dataset_test)
output_config = {
                'steps_per_epoch': steps_per_epoch,
                'callbacks': [
                    ModelSaver(max_to_keep=10, keep_checkpoint_every_n_hours=0.33),
                    MaxSaver('MAF_loss', filename='max_log_prob'),
                    LogProbAfterEpoch("EstimatePDF/MAF_loss"),
                    InferenceRunner(
                        dataset_test,
                        [
                            LogProbSum("EstimatePDF/MAF_loss")
                        ]
                    ),
                    ScheduledHyperParamSetter('optimizer/learning_rate', [
                        (a.pdf_estimator['drop0'], a.pdf_estimator['lr0']),
                        (a.pdf_estimator['drop1'], a.pdf_estimator['lr1'])
                    ])
                ],
                'max_epoch': a.pdf_estimator['hm_epoch']
        }
```

4. here is code of collections:

```
        analyze_variables = tf.get_collection(
            tf.GraphKeys.TRAINABLE_VARIABLES,
            scope='.*analyze'
        )
        pdf_variables = tf.get_collection(
            tf.GraphKeys.TRAINABLE_VARIABLES,
            scope='.*EstimatePDF'
        )
        train_variables = tf.get_collection(
            tf.GraphKeys.TRAINABLE_VARIABLES,
            scope='^(?!(.*(EstimatePDF|analyze))).*$'
        )
        databus['analyze_variables'] = analyze_variables
        databus['pdf_variables'] = pdf_variables
        databus['train_variables'] = train_variables

        G = tf.get_default_graph()
        for tensor in databus['analyze_variables']:
            G.add_to_collection(name='analyze_variables', value=tensor)
        for tensor in databus['pdf_variables']:
            G.add_to_collection(name='pdf_variables', value=tensor)
        for tensor in databus['train_variables']:
            G.add_to_collection(name='train_variables', value=tensor)
```

I don't think I put any non variable into the collections.


here are callbacks I used:

```
class AfterTrainEpoch(Callback):
    def _setup_graph(self):
        self.train_tower = self.trainer.tower_func.towers[0]
        self.outputs = [self.train_tower.get_tensor(name) for name in self.get_fetches()]
    def _before_train(self):
        pass
    def _before_epoch(self):
        self._before_train_epoch()
    def _before_train_epoch(self):
        pass
    def _before_run(self, _):
        return tf.train.SessionRunArgs(fetches=self.outputs)
    def _after_run(self, _, run_values):
        self.on_fetches(run_values.results)
    def on_fetches(self, outputs):
        self._on_fetches(outputs)
    def _on_fetches(self, outputs):
        raise NotImplementedError()
    def _trigger_epoch(self):
        ret = self._after_train_epoch()
        if ret is None:
            return
        for k, v in six.iteritems(ret):
            try:
                v = float(v)
            except ValueError:
                logger.warn("{} returns a non-scalar statistics!".format(type(self).__name__))
                continue
            else:
                self.trainer.monitors.put_scalar(k, v)
    def _after_train_epoch(self):
        pass
    def get_fetches(self):
        try:
            ret = self._get_fetches()
        except NotImplementedError:
            logger.warn("Inferencer._get_output_tensors was deprecated and renamed to _get_fetches")
            ret = self._get_output_tensors()
        return [get_op_tensor_name(n)[1] for n in ret]
    def _get_fetches(self):
        raise NotImplementedError()
    def _get_output_tensors(self):
        pass
```

```
class LogProbAfterEpoch(AfterTrainEpoch):
    def __init__(self, log_prob_tensor_name):
        """
        Args:
            log_prob_tensor_name(str): name of the LogProb tensor.
        """
        self.log_prob_tensor_name = log_prob_tensor_name
    def _before_train_epoch(self):
        self.collector = list()
    def _get_fetches(self):
        return [self.log_prob_tensor_name]
    def _on_fetches(self, outputs):
        log_prob = outputs[0]
        self.collector.append(log_prob)
    def _after_train_epoch(self):
        log_prob = np.concatenate(self.collector, axis=0)
        return {
            'train_MAF_loss': np.sum(log_prob)
        }
```

```
class LogProbSum(Inferencer):
    def __init__(self, log_prob_tensor_name):
        """
        Args:
            log_prob_tensor_name(str): name of the LogProb tensor.
        """
        self.log_prob_tensor_name = log_prob_tensor_name
    def _before_inference(self):
        self.collector = list()
    def _get_fetches(self):
        return [self.log_prob_tensor_name]
    def _on_fetches(self, outputs):
        log_prob = outputs[0]
        self.collector.append(log_prob)
    def _after_inference(self):
        log_prob = np.concatenate(self.collector, axis=0)
        return {
            'MAF_loss': np.sum(log_prob)
        }
```

**Comment by mikeun** [comment]
Also I found this issue:  [https://github.com/tensorflow/tensorflow/issues/19657](url)
about the same problem.

**Comment by ppwwyyxx** [comment]
OK. Looks like that's a tensorflow bug then. Seems to happen when you have a very large constant in the graph.

**Comment by mikeun** [comment]
Does it mean - I have value ~ 2**32 in my dataflow?
I made all my data normalized from 0 to 1 and saved it to TFrecords with compression.
Dataflow:
```
        train_size, test_size = get_train_test_size()

        train = TFRecordData(
            a.folders['inp'] + a.tf_records['train_file'],
            size=train_size,
            compress=True
        )
        train = BatchData(
            train,
            a.start_args['train_batch_size'],
            remainder=True)

        test = TFRecordData(
            a.folders['inp'] + a.tf_records['test_file'],
            size=test_size,
            compress=True
        )
        test = BatchData(
            test,
            a.start_args['test_batch_size'],
            remainder=True)
```

**Comment by ppwwyyxx** [comment]
No I think it's caused by large constant tensor in the graph -- not necessarily in the data part, could be anywhere in the graph.
But that's just my guess and you'd better wait for responses from the TF team.

**Comment by mikeun** [comment]
I don't understand what does it mean exactly?
```
[0606 10:50:47 @collection.py:151] Size of these collections were changed in InferenceTower: (analyze_variables: 1->2), (pdf_variables: 72->144), (train_variables: 174->348)
[0606 10:50:47 @collection.py:164] These collections were modified but restored in InferenceTower: (tf.GraphKeys.UPDATE_OPS: 8->16)
```
why it changed collections size (*2) during InferenceTower?

**Comment by mikeun** [comment]
I reduced number of parameters and the model goes on. Maybe there is number of parameters limit.

**Comment by ppwwyyxx** [comment]
See http://tensorpack.readthedocs.io/tutorial/trainer.html#tower-trainer 
Your model function will be called multiple times, and as a result will change your collection.

**Comment by ppwwyyxx** [comment]
I'll close this since this appears to be a tensorflow bug https://github.com/tensorflow/tensorflow/issues/19657 that a graph cannot be too large.
